lxd/storage/drivers/driver_btrfs_utils.go

```diff
@@ -107,39 +108,70 @@ func (d *btrfs) hasSubvolumes(path string) (bool, error) {
 }
 
 func (d *btrfs) getSubvolumes(path string) ([]string, error) {
+	// Make sure the path has a trailing slash.
+	if !strings.HasSuffix(path, "/") {
+		path = path + "/"
+	}
+
 	poolMountPath := GetPoolMountPath(d.name)
 	if !strings.HasPrefix(path, poolMountPath+"/") {
 		return nil, fmt.Errorf("%q is outside pool mount path %q", path, poolMountPath)
 	}
 
-	path = strings.TrimPrefix(path, poolMountPath+"/")
+	var result []string
 
-	// Make sure the path has a trailing slash.
-	if !strings.HasSuffix(path, "/") {
-		path = path + "/"
-	}
+	if !d.state.OS.RunningInUserNS {
```

Comment:
This feels a bit "mixed up" now to me.
We have the comment about the permission error inside the stanza that is used for when containers arent being nested.
And the comment about "revert to walking..." doesn't mention why it is using this approach.

How about we swap this around and do something like:

```go
if d.state.OS.RunningInUserNS {
    // If using BTRFS in a nested container we cannot use "btrfs subvolume list" due to a permission error.
    // So instead walk the directory tree testing each directory to see if it is subvolume.
    err := filepath.Walk(path, func(fpath string, entry fs.FileInfo, err error) error {
    ...
} else {
    // If not running inside a nested container we can use "btrfs subvolume list" to get subvolumes which is
    // more performant than walking the directory tree.
    var stdout bytes.Buffer
    err := shared.RunCommandWithFds(d.state.ShutdownCtx, nil, &stdout, "btrfs", "subvolume", "list", poolMountPath)
    ...
}
```

Reply:
Ok looking clearer now thanks :)